lib/carelink/bolusNWizard.js

```diff
@@ -470,129 +472,164 @@ module.exports = function(timezone, model) {
   }
 
   var datumsBySeqNum = null, findNormal = null, findSquare = null, findWizard = null;
+  var pointer = 0, cachedObjects = null;
 
-  return function(simulator, datum, i, data) {
-    var normal = null, square = null, wizard = null;
-    var parsed = parser(datum);
+  function connectEvents(data) {
+    cachedObjects = [];
+    for (var i = 0; i < data.length; ++i) {
+      var normal = null, square = null, wizard = null;
+      var parsed = parser(data[i]);
 
-    if (datumsBySeqNum == null) {
-      buildCache(data);
-    }
+      if (datumsBySeqNum == null) {
+        buildCache(data);
+      }
 
-    if (findNormal == null) {
-      findNormal = makeFindNormalFn();
-    }
+      if (findNormal == null) {
+        findNormal = makeFindNormalFn();
+      }
 
-    if (findSquare == null) {
-      findSquare = makeFindSquareFn();
-    }
+      if (findSquare == null) {
+        findSquare = makeFindSquareFn();
+      }
 
-    if (findWizard == null) {
-      findWizard = makeFindWizardFn();
-    }
-    
-    if (parsed != null) {
-      switch(parsed.type) {
-        case 'bolusNormal':
-          if (isInCache(parsed)) {
-            removeFromCache(parsed);
-            normal = cleanEvent(parsed);
-            if (parsed.dualComponent) {
-              square = findSquare(parsed);
-              if (square != null) {
-                datumsBySeqNum[square.uploadId][square.uploadSeqNum] = null;
+      if (findWizard == null) {
+        findWizard = makeFindWizardFn();
+      }
+      
+      if (parsed != null) {
+        switch(parsed.type) {
+          case 'bolusNormal':
+            if (isInCache(parsed)) {
```

Comment:
In response to first question: no. The problem that the cache is solving is that we sometimes need to process events *before* we get to them in order of timestamp/rows in the CSV. For example, we often come across a wizard, look for an associated bolus...if we find such a bolus, we process both and remove *both* wizard and bolus from the cache. Then the next row in the CSV to be processed might very well be the bolus we just removed from the cache. If it's no longer in the cache, we just skip to the next thing because we know we already processed it.

The test for `parsed` is a little different. I'd call this one a cheddarism - i.e., the pattern of initializing to null and then checking later to see if something is still null. If you look in the CareLink driver [here](https://github.com/tidepool-org/chrome-uploader/blob/jebeck/bolus-and-wiz-seq/lib/drivers/carelinkDriver.js#L51) and [here](https://github.com/tidepool-org/chrome-uploader/blob/jebeck/bolus-and-wiz-seq/lib/drivers/carelinkDriver.js#L233), you'll see that each row of the CSV gets passed to *every* processor. And the pattern for all the processors is to start with `parsed = null` then try to parse the CSV row with the local row-parser (built using `common.makeParser`), and each local row-parser only attempts to parse rows of certain types (whatever's in the domain of that processor) and returns `null` otherwise.

Does that help?